File: Sorting/HeapSort.py
import random
import time

n = 40000000
heap = []
heap = heap + [-1]*(n+1)
srtd = []
root = 0
last = 0
size = 0

def left(root):
    return root*2+1

def right(root):
    return root*2+2
    
def parent(root):
    return (root-1)/2
    
def insert(data):
    global last
    global size
    heap[last] = data
    shiftUp(last)
    last = last+1
    size = size+1

def swap(i, j):
    tmp = heap[i]
    heap[i] = heap[j]
    heap[j] = tmp
    
def shiftUp(pos):
    while heap[pos] > heap[parent(pos)]:
        swap(pos, parent(pos))
        pos = parent(pos)
        
def shiftDown(pos):
    bigger = 0
    if right(pos) <= last and heap[right(pos)] > heap[left(pos)]:
        bigger = right(pos)
    else:
        bigger = left(pos)
        
    if bigger <= last and heap[pos] < heap[bigger]:
        swap(bigger, pos)
        pos = bigger
        shiftDown(pos)
        
def deleteRoot():
    global root
    tmp = heap[root]
    delete(root)
    return tmp
    
def delete(pos):
    global last
    global size
    heap[pos] = heap[last]
    heap[last] = -1
    last = last-1
    shiftUp(pos)
    shiftDown(pos)
    size = size-1
    
def HeapSort():
    while last != 0:
        gg = deleteRoot()
        srtd.append(gg)


for i in range(0, n):
    ins = random.randint(0, 1000)
    insert(ins)

start = time.clock()
HeapSort()
end = time.clock()
print ("MergeSort time : " + str(end-start) + " detik")    

File: Sorting/HeapSort.cpp
#include <bits/stdc++.h>
using namespace std;

long *heap, *sorted;
int root, last, size;

    int left(int root){
        return root*2+1;
    }

    int right(int root){
        return root*2+2;
    }

    int parent(int root){
        return (root-1)/2;
    }

    void shiftUp(int tmp){
        while(heap[tmp] > heap[parent(tmp)]){
            swap(heap[tmp], heap[parent(tmp)]);
            tmp = parent(tmp);
        }
    }

    void insert(long data){
        heap[last] = data;
        shiftUp(last);
        last++;
        size++;
    }

    void shiftDown(int tmp){
        int bigger;
        if(right(tmp) <= last && heap[right(tmp)] > heap[left(tmp)]){
            bigger = right(tmp);
        }
        else bigger = left(tmp);

        if(bigger <= last && heap[tmp] < heap[bigger]){
            swap(heap[tmp], heap[bigger]);
            tmp = bigger;
            shiftDown(tmp);
        }
    }

    void dlt(int pos){
        heap[pos] = heap[last];
        last--;
        heap[last+1] = -1;
        shiftUp(pos);
        shiftDown(pos);
        size--;
    }

    long deleteRoot(){
        long tmp = heap[root];
        dlt(root);
        return tmp;
    }

    void sort(){
        int idx = 0;
        while(last != 0){
            long gg = deleteRoot();
            //System.out.println(gg + " ");
            sorted[idx] = gg;
            idx++;
        }
    }

int main(){
    int n = 160000000;
    heap = new long[n+1000];
    sorted = new long[n+100];
    root = last = size = 0;
    for (int i = 0; i < n; i++) {
        insert(rand()*1000000000+1);
    }

    double start = clock();
    sort();
    cout << "Waktu HeapSort : " << fixed << double(clock()-start)/CLOCKS_PER_SEC << " detik" << endl;

}
